### PR TITLE
Hclang 123 examples

### PR DESCRIPTION
**Pull request by C0DK**
### Description
Added a lot of tests and fixed all the bugs raised by them.

### Issues solved 
add these with the respective key as HCLANG-1 (as an example)
 - HCLANG-123

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding code [if applicable]
- Check [coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)
- Are public methods documented? 
- Are there testcases for new features / methods?
- Have you tried running the source and tested things that cannot be test-cased?

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

File: code/src/test/kotlin/hclCodeTests/TestHclPrograms.kt [hunk 1]
```diff
@@ -28,12 +26,24 @@ fun generateFilesFromCode(code: String): List<FilePair> {
...
+        "HelloWorldAndReturn.hcl",
+        "multiScope.hcl",
+        "printTuple.hcl",
         "stringAt.hcl",
-        "printTuple.hcl"
+        "Swap.hcl",
+        "printTuple.hcl",
+        "OOP.hcl",
+        "OOP_V2.hcl",
+            "Swap.hcl"
```
Comment (nillas12) [inline code/src/test/kotlin/hclCodeTests/TestHclPrograms.kt]: Indent?

File: code/src/test/resources/PrintFibonacci.hcl [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# This should print [1, 1, 2, 3, 5, 8]
+
+func fib = (num n) : num {
+    var val = 0
+    (n lessThanEqual 2) then { val = 1 }
+    (n lessThanEqual 2) not then { val = ((n - 1) fib) + ((n - 2) fib) }
+    return val
+}
+
+1 to 6 map { value fib } print
```
Comment (nillas12) [inline code/src/test/resources/PrintFibonacci.hcl]: No newline

File: code/src/test/resources/fizzBuzz.hcl [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	(v mod 5 equals 0) then { toPrint = toPrint + "buzz" }
+    toPrint equals "" then { toPrint = v toText }
+    return toPrint
+
+
+}
+
+1 to 15 map { value fizzbuzz } print
+
+RETURN_CODE = 0
```
Comment (nillas12) [inline code/src/test/resources/fizzBuzz.hcl]: No fileend

File: code/src/test/resources/subText.hcl [hunk 4]
```diff
@@ -1,2 +1,2 @@
 # should print hello
-"hello world" subText 0 6
\ No newline at end of file
+"hello world" subText 0 5 print
```
Comment (nillas12) [inline code/src/test/resources/subText.hcl]: No endfile

File: code/src/test/resources/thenElse.hcl [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    T output = trueBody
+    cond not then { output = falseBody }
+    return output
+}
+
+var x = 21
+var y = 20
+num bigNum
+bigNum = x greaterThan y thenElse { x } { y }
+RETURN_CODE = bigNum
```
Comment (nillas12) [inline code/src/test/resources/thenElse.hcl]: No endfile

File: code/src/test/resources/toUneven.hcl [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+var first = (list[T] myList): T {
+    myList at 0
+}
+
+var numbers = 1 toUneven 10
+
+
+"first=" + (numbers first toText) + " and last=" + (numbers last toText) print
```
Comment (nillas12) [inline code/src/test/resources/toUneven.hcl]: No endfile

File: code/src/main/kotlin/generation/cpp/HelperHeaders.kt [hunk 7]
```diff
@@ -118,14 +118,18 @@ List<char> ftoa(double d, int precision) {
 
             mantissa -= characteristic;
         }
-        while(*(endOfString-1) == '0' || *(endOfString-1) == '.'){
+        while(*(endOfString-1) == '0'){
+            endOfString = endOfString - 1;
+        }
+        while(*(endOfString-1) == '.'){
```
Comment (Lambcreek) [inline code/src/main/kotlin/generation/cpp/HelperHeaders.kt]: Der kan kun være 1 ' . ' i strengen, så der er ikke brug for en while løkke

**Pull request merged**